crawl-ref/source/newgame.cc

```diff
@@ -182,32 +182,37 @@ static void _resolve_species(newgame_def& ng, const newgame_def& ng_choice)
         return;
 
     case SP_VIABLE:
-        // If we have a job already picked out, pick a species which recommends
-        // it. Otherwise, pick a random species.
+        // If we have already picked a job, find a species that recommends it
         if (is_starting_job(ng.job))
         {
-            do
-            {
-                ng.species = random_species();
-            } while (!job_recommends_species(ng.job, ng.species));
+            auto species_recommended_by_job =
+                filter_enum(species_type, [&](species_type species) {
+                    return job_recommends_species(ng.job, species);
+                });
+            // Note: we know the array has at least one element because all
```

Comment:
Might want to `ASSERT(species_recommended_by_job.size() > 0)` here; `random2` has the unfortunate property that a range of size 0 will always return 0, instead of faulting from div-0. Even if it's guaranteed by other checks, defense in depth never hurts.